litedram/phy/lpddr5/commands.py

```diff
@@ -250,40 +254,49 @@ def set(self, cmd):
-            "SB(\d+)": lambda i: 0,  # sub-bank selection related to RFM
-            "WS_WR":   lambda: self.wck_sync == WCKSyncType.WR,  # Write WCK2CK SYNC
-            "WS_RD":   lambda: self.wck_sync == WCKSyncType.RD,  # Read WCK2CK SYNC
-            "WS_FS":   lambda: self.wck_sync == WCKSyncType.FS,  # FAST SYNC
-            "DC(\d+)": lambda i: 0,  # Data Copy, unimplemented
-            "WRX":     lambda: 0,  # Write X function, unimplemented
-            "WXSA":    lambda: 0,  # Write X function, unimplemented
-            "WXSB":    lambda: 0,  # Write X function, unimplemented
-            "BA(\d+)": lambda i: self.dfi.bank[i],  # only BA0-2 is used, in BG/B16 modes we always refresh banks (x, x+8)
-            "R(\d+)":  lambda i: self.dfi.address[i],  # row
+            "H":         lambda: 1,  # high
+            "L":         lambda: 0,  # low
+            "V":         lambda: 0,  # defined logic
+            "X":         lambda: 0,  # don't care
+            "AB":        lambda: self.dfi.address[10],  # all banks
+            "AP":        lambda: self.dfi.address[10],  # auto precharge
+            "RFM":       lambda: 0,  # TODO: 1=RFM, 0=REF (Refresh Managemenent, only if r/o MR[27][0]=1, else always REF)
+            "SB(\d+)":   lambda i: 0,  # sub-bank selection related to RFM
+            "WS_WR":     lambda: self.wck_sync == WCKSyncType.WR,  # Write WCK2CK SYNC
+            "WS_RD":     lambda: self.wck_sync == WCKSyncType.RD,  # Read WCK2CK SYNC
+            "WS_FS":     lambda: self.wck_sync == WCKSyncType.FS,  # FAST SYNC
+            "DC(\d+)":   lambda i: 0,  # Data Copy, unimplemented
+            "WRX":       lambda: 0,  # Write X function, unimplemented
+            "WXSA":      lambda: 0,  # Write X function, unimplemented
+            "WXSB":      lambda: 0,  # Write X function, unimplemented
+            "BA(\d+)":   lambda i: self.dfi.bank[i],  # only BA0-2 is used, in BG/B16 modes we always refresh banks (x, x+8)
+            "R(\d+)":    lambda i: self.dfi.address[i],  # row
             # LPDDR5 specs split the regular column address into C[5:0] "column address" and B[3:0] "burst address"
-            "C(\d+)":  lambda i: self.dfi.address[i + 4],
-            "MA(\d+)": lambda i: mr_address[i],  # mode register address
-            "OP(\d+)": lambda i: self.dfi.address[i],  # mode register value, or operand for MPC
+            "C(\d+)":    lambda i: self.dfi.address[i + 4],
+            "MA(\d+)":   lambda i: mr_address[i],  # mode register address
+            "OP(\d+)":   lambda i: self.dfi.address[i],  # mode register value, or operand for MPC
+            # FIXME: The LiteDRAM refresher sends a ZQC command without updating the addresses.
+            # This is a workaround that adds an additional fake operand when a ZQC is called by the LiteDRAM refresher.
+            "ZQOP(\d+)": lambda i: int("{:b}".format(mpc_fixed_address)[7-i]),
```

Comment:
I wonder if it would be possible to avoid modifying the TRUTH_TABLE (so that it is the same as in specification), but I am not sure if might be missing something. Would the following work?
*  reverting ZQOP and changes related to MPC-ZQCS/MPC-ZQCL
* keep the `cmd_str` argument
* in  `Command.parse_bit` have the following:
```python
is_mpc = cmd_str == "MPC"
# FIXME: this is a workaround for Refresher sending ZQC as MPC with address=0
op = Constant(MPC.ZQC_LATCH, 8) if is_mpc and self.dfi.address == 0 else self.dfi.address
# ...
"OP(\d+)": lambda i: op[i],  # mode register value, or operand for MPC
```
I haven't tested it though, what do you think about it? The main motivation is that it should be easier to get rid of this workaround later when we make the Refresher configurable.

Reply:
I have actually tested a very similar approach (as well as these changes), but unfortunately this does not seem to be working correctly. Mainly, I think the problem is that we would be mixing migen with normal python, and the `self.dfi.address == 0` will be always false in this case, letting the `else` branch to be always selected, despite the command is an `MPC`

Reply:
With the suggested change I get the following errors during simulation:

```
000[       183863750 ps] [WARN] Only MRW/MRR commands expected before ZQ Latch ...
[       183863750 ps] [WARN] ... ACT=0 PRE=1 REF=0 MRW=0 DATA=0 CAS=0 MPC=0
[       183893750 ps] [WARN] Only MRW/MRR commands expected before ZQ Latch ...
[       183893750 ps] [WARN] ... ACT=0 PRE=0 REF=1 MRW=0 DATA=0 CAS=0 MPC=0
0[       184113750 ps] [WARN] Only MRW/MRR commands expected before ZQ Latch ...
[       184113750 ps] [WARN] ... ACT=0 PRE=1 REF=0 MRW=0 DATA=0 CAS=0 MPC=0
[       184143750 ps] [WARN] Unsupported MPC command: op=0b00000000
[       184143750 ps] [ERROR] ZQC-LATCH expected, got MPC with op=0b00000000
...[       187773750 ps] [WARN] Only MRW/MRR commands expected before ZQ Latch ...
[       187773750 ps] [WARN] ... ACT=0 PRE=1 REF=0 MRW=0 DATA=0 CAS=0 MPC=0
[       187803750 ps] [WARN] Only MRW/MRR commands expected before ZQ Latch ...
[       187803750 ps] [WARN] ... ACT=0 PRE=0 REF=1 MRW=0 DATA=0 CAS=0 MPC=0
[       188023750 ps] [WARN] Only MRW/MRR commands expected before ZQ Latch ...
[       188023750 ps] [WARN] ... ACT=0 PRE=1 REF=0 MRW=0 DATA=0 CAS=0 MPC=0
[       188053750 ps] [WARN] Unsupported MPC command: op=0b00000000
[       188053750 ps] [ERROR] ZQC-LATCH expected, got MPC with op=0b00000000

```

Reply:
Ah, you're totally right, it's mixing regular Python and Migen, my bad. This should probably work:
```python
mpc_op = Signal(8)
self.comb += If(self.dfi.address == 0,
    mpc_op.eq(MPC.ZQC_LATCH)
).Else(
    mpc_op.eq(self.dfi.address)
)
op = mpc_op if is_mpc else self.dfi_address
```

Reply:
All right, thanks, this works. I have removed the truth table changes